### PR TITLE
[GSoC] Added Log-logistic distribution

### PR DESCRIPTION
**Pull request by ritesh99rakesh**
#### Brief description of what is fixed or changed
Added log-logistic distribution under `sympy/stats/crv_types.py`
Also added corresponding tests.

#### Release Notes

- stats
    - Added Log-logistic distribution in `sympy/stats/crv_types.py`

### CODE REVIEW

**Comment by ritesh99rakesh** [comment]
@sidhantnagpal Please review this.

File: sympy/stats/crv_types.py [hunk 1]
```diff
@@ -1917,6 +1919,95 @@ def Logistic(name, mu, s):
...
+    set = Interval(0, oo)
+
+    @staticmethod
+    def check(alpha, beta):
+        _value_check(alpha > 0, "Scale parameter Alpha must be positive.")
+        _value_check(beta > 0, "Shape parameter Beta must be positive.")
+
+    def pdf(self, x):
+        a, b = self.alpha, self.beta
+        return ((b / a) * (x / a)**(b - 1)) / (1 + (x / a)**b)**2
```
Comment (sidhantnagpal) [inline sympy/stats/crv_types.py]: Spaces are generally included around operators `+` and `-` to distinguish the terms.
The spaces within terms (ie around `*` or `/`) can be skipped.

**Comment by sidhantnagpal** [comment]
Did you look into why `E(X)` is not able to produce a simpler expression for mean? I get:
```
In [8]: E(X)                                                                                                               
Out[8]: Piecewise((pi*(1/a)**(b - 1)*polar_lift(a)**b*polar_lift(a)**(b*(-1 + (b + 1)/b))/(a*sin(pi/b)*Abs(b)), ((Abs(log(Abs(a))*im(b) + re(b)*arg(a)) < pi) & ((re(b) + 1)*re(b)/(re(b)**2 + im(b)**2) - 1 + im(b)**2/(re(b)**2 + im(b)**2) > -1) & ((re(b) + 1)*re(b)/(re(b)**2 + im(b)**2) - 3 + im(b)**2/(re(b)**2 + im(b)**2) < -1)) | (Ne(polar_lift(a)**(-b), 0) & (Abs(log(Abs(a))*im(b) + re(b)*arg(a)) < pi) & Ne(Abs(log(Abs(a))*im(b) + re(b)*arg(a)), pi) & ((re(b) + 1)*re(b)/(re(b)**2 + im(b)**2) - 1 + im(b)**2/(re(b)**2 + im(b)**2) > -1) & ((re(b) + 1)*re(b)/(re(b)**2 + im(b)**2) - 3 + im(b)**2/(re(b)**2 + im(b)**2) < -1))), (Piecewise((pi*(1/a)**(b - 1)*polar_lift(a)**b*polar_lift(a)**(b*(-1 + (b + 1)/b))/(a*sin(pi/b)*Abs(b)), ((Abs(log(Abs(a))*im(b) + re(b)*arg(a)) < pi) & ((re(b) + 1)*re(b)/(re(b)**2 + im(b)**2) - 1 + im(b)**2/(re(b)**2 + im(b)**2) > -1) & ((re(b) + 1)*re(b)/(re(b)**2 + im(b)**2) - 3 + im(b)**2/(re(b)**2 + im(b)**2) < -1)) | (Ne(polar_lift(a)**(-b), 0) & (Abs(log(Abs(a))*im(b) + re(b)*arg(a)) < pi) & Ne(Abs(log(Abs(a))*im(b) + re(b)*arg(a)), pi) & ((re(b) + 1)*re(b)/(re(b)**2 + im(b)**2) - 1 + im(b)**2/(re(b)**2 + im(b)**2) > -1) & ((re(b) + 1)*re(b)/(re(b)**2 + im(b)**2) - 3 + im(b)**2/(re(b)**2 + im(b)**2) < -1))), (Integral(X*b*(X/a)**(b - 1)/(a*((X/a)**b + 1)**2), (X, 0, oo)), True)), True))
```
It should be able to give something like `pi*a/(b*sin(pi/b))`.

**Comment by ritesh99rakesh** [comment]
> Did you look into why `E(X)` is not able to produce a simpler expression for mean? I get:
> 
> ```
> In [8]: E(X)                                                                                                               
> Out[8]: Piecewise((pi*(1/a)**(b - 1)*polar_lift(a)**b*polar_lift(a)**(b*(-1 + (b + 1)/b))/(a*sin(pi/b)*Abs(b)), ((Abs(log(Abs(a))*im(b) + re(b)*arg(a)) < pi) & ((re(b) + 1)*re(b)/(re(b)**2 + im(b)**2) - 1 + im(b)**2/(re(b)**2 + im(b)**2) > -1) & ((re(b) + 1)*re(b)/(re(b)**2 + im(b)**2) - 3 + im(b)**2/(re(b)**2 + im(b)**2) < -1)) | (Ne(polar_lift(a)**(-b), 0) & (Abs(log(Abs(a))*im(b) + re(b)*arg(a)) < pi) & Ne(Abs(log(Abs(a))*im(b) + re(b)*arg(a)), pi) & ((re(b) + 1)*re(b)/(re(b)**2 + im(b)**2) - 1 + im(b)**2/(re(b)**2 + im(b)**2) > -1) & ((re(b) + 1)*re(b)/(re(b)**2 + im(b)**2) - 3 + im(b)**2/(re(b)**2 + im(b)**2) < -1))), (Piecewise((pi*(1/a)**(b - 1)*polar_lift(a)**b*polar_lift(a)**(b*(-1 + (b + 1)/b))/(a*sin(pi/b)*Abs(b)), ((Abs(log(Abs(a))*im(b) + re(b)*arg(a)) < pi) & ((re(b) + 1)*re(b)/(re(b)**2 + im(b)**2) - 1 + im(b)**2/(re(b)**2 + im(b)**2) > -1) & ((re(b) + 1)*re(b)/(re(b)**2 + im(b)**2) - 3 + im(b)**2/(re(b)**2 + im(b)**2) < -1)) | (Ne(polar_lift(a)**(-b), 0) & (Abs(log(Abs(a))*im(b) + re(b)*arg(a)) < pi) & Ne(Abs(log(Abs(a))*im(b) + re(b)*arg(a)), pi) & ((re(b) + 1)*re(b)/(re(b)**2 + im(b)**2) - 1 + im(b)**2/(re(b)**2 + im(b)**2) > -1) & ((re(b) + 1)*re(b)/(re(b)**2 + im(b)**2) - 3 + im(b)**2/(re(b)**2 + im(b)**2) < -1))), (Integral(X*b*(X/a)**(b - 1)/(a*((X/a)**b + 1)**2), (X, 0, oo)), True)), True))
> ```
> 
> It should be able to give something like `pi*a/(b*sin(pi/b))`.

This is actually correct. When you look at the [wiki](https://en.wikipedia.org/wiki/Log-logistic_distribution), it defines `mean` only when `beta > 1`. Following works:
```
>>> from sympy.stats import *
>>> from sympy import symbols
>>> a = symbols('a', positive=True)
>>> b = symbols('b', prime=True)
>>> b > 1
True
>>> X = LogLogistic('x', a, b)
>>> E(X).expand()
pi*a/(b*sin(pi/b))
```
Do you know how I can put explicit condition that `b > 1` in `symbols`. I will add a test for this.

**Comment by sidhantnagpal** [comment]
The _last comment_ had no assumptions set for `a` and `b`.
It gets slightly better with `a, b = symbols('a b', positive=True)`:
```
In [13]: X = LogLogistic('X', a, b)                                             

In [14]: E(X)                                                                   
Out[14]: Piecewise((pi*a**b*a**(-b*(1 - (b + 1)/b))*a**(1 - b)/(a*b*sin(pi/b)), 3 - (b + 1)/b > 1), (Piecewise((pi*a**b*a**(-b*(1 - (b + 1)/b))*a**(1 - b)/(a*b*sin(pi/b)), 3 - (b + 1)/b > 1), (Integral(X*b*(X/a)**(b - 1)/(a*((X/a)**b + 1)**2), (X, 0, oo)), True)), True))
```

> This is actually correct

It should ideally be undefined for `b <= 1`.

> Do you know how I can put explicit condition that `b > 1` in `symbols`

I do not know if that is possible. One possible workaround is:
```
In [9]: c = Symbol('c', positive=True)                                          

In [10]: (1 + c) > 1                                                            
Out[10]: True
```
But SymPy is unable to finish computing `E(X)` for `X = LogLogistic('X', a, c + 1)`.

**Comment by czgdp1807** [comment]
> I do not know if that is possible. One possible workaround is:

As far as I know, currently the work is under progress for such assumptions. May be @asmeurer can throw more light on this.



> Did you look into why `E(X)` is not able to produce a simpler expression for mean? I get:
> 
> ```
> In [8]: E(X)                                                                                                               
> Out[8]: Piecewise((pi*(1/a)**(b - 1)*polar_lift(a)**b*polar_lift(a)**(b*(-1 + (b + 1)/b))/(a*sin(pi/b)*Abs(b)), ((Abs(log(Abs(a))*im(b) + re(b)*arg(a)) < pi) & ((re(b) + 1)*re(b)/(re(b)**2 + im(b)**2) - 1 + im(b)**2/(re(b)**2 + im(b)**2) > -1) & ((re(b) + 1)*re(b)/(re(b)**2 + im(b)**2) - 3 + im(b)**2/(re(b)**2 + im(b)**2) < -1)) | (Ne(polar_lift(a)**(-b), 0) & (Abs(log(Abs(a))*im(b) + re(b)*arg(a)) < pi) & Ne(Abs(log(Abs(a))*im(b) + re(b)*arg(a)), pi) & ((re(b) + 1)*re(b)/(re(b)**2 + im(b)**2) - 1 + im(b)**2/(re(b)**2 + im(b)**2) > -1) & ((re(b) + 1)*re(b)/(re(b)**2 + im(b)**2) - 3 + im(b)**2/(re(b)**2 + im(b)**2) < -1))), (Piecewise((pi*(1/a)**(b - 1)*polar_lift(a)**b*polar_lift(a)**(b*(-1 + (b + 1)/b))/(a*sin(pi/b)*Abs(b)), ((Abs(log(Abs(a))*im(b) + re(b)*arg(a)) < pi) & ((re(b) + 1)*re(b)/(re(b)**2 + im(b)**2) - 1 + im(b)**2/(re(b)**2 + im(b)**2) > -1) & ((re(b) + 1)*re(b)/(re(b)**2 + im(b)**2) - 3 + im(b)**2/(re(b)**2 + im(b)**2) < -1)) | (Ne(polar_lift(a)**(-b), 0) & (Abs(log(Abs(a))*im(b) + re(b)*arg(a)) < pi) & Ne(Abs(log(Abs(a))*im(b) + re(b)*arg(a)), pi) & ((re(b) + 1)*re(b)/(re(b)**2 + im(b)**2) - 1 + im(b)**2/(re(b)**2 + im(b)**2) > -1) & ((re(b) + 1)*re(b)/(re(b)**2 + im(b)**2) - 3 + im(b)**2/(re(b)**2 + im(b)**2) < -1))), (Integral(X*b*(X/a)**(b - 1)/(a*((X/a)**b + 1)**2), (X, 0, oo)), True)), True))
> ```
> 
> It should be able to give something like `pi*a/(b*sin(pi/b))`.

My personal opinion is to let it be the result complicated for now. I will be working on simplifying the results of stats module in phase 3 so we will discuss the issue in detail there. I will reference this example. 
We can add numeric tests for now for E(X).

File: sympy/stats/crv_types.py (hunk 1 above)
Comment (Abdullahjavednesar) [reply sympy/stats/crv_types.py]: go through [this](https://www.python.org/dev/peps/pep-0008/#whitespace-in-expressions-and-statements).

**Comment by ritesh99rakesh** [comment]
@sidhantnagpal Please tell me what changes have to be made to this PR.
I tried the following, but it is taking forever to compute:
```
>>> a = symbols('a', positive=True)
>>> b = 1 + symbols('b', positive=True)
>>> X = LogLogistic('x', a, b)
>>> E(X) # takes forever to compute
```
What might be the reason for this? @asmeurer please guide me here, then I will work on making it work.

**Comment by jksuom** [comment]
It seems that `meijerint` can compute the integral if the parameter `b` is a Symbol but not if it is `b + 1`. (It probably cannot make use of the condition `b > 0` in the latter case.) It follows that `heurisch` will be tried, and that may really take what looks like forever. As a workaround (until `meijerint` is fixed) it is possible to use `E(X, heurisch=False)`. (That should currently return an unevaluated integral in a reasonable time.)

**Comment by jksuom** [comment]
`meijerint` tries to match the factor `1/((x/a)**(b + 1) + 1)**2` of the pdf to the pattern `(x**q_*p_ + b_)**(-a_)`. It succeeds when the exponent is `b` but fails with `b + 1` because `(x/a)**(b + 1)/a**(-b - 1)` does not simplify:
```
>>> x**(b + 1)*a**(-b - 1)/a**(-b - 1)
a**(-b - 1)*a**(b + 1)*x**(b + 1)
>>> x**b*a**-b/a**-b
x**b
```

**Comment by jksuom** [comment]
This is one way to make `match` succeed:
```
--- a/sympy/core/mul.py
+++ b/sympy/core/mul.py
@@ -1032,6 +1032,7 @@ def _combine_inverse(lhs, rhs):
         Returns lhs/rhs, but treats arguments like symbols, so things like
         oo/oo return 1, instead of a nan.
         """
+        from sympy.simplify.powsimp import powsimp
         if lhs == rhs:
             return S.One
 
@@ -1056,7 +1057,7 @@ def check(l, r):
                 else:
                     b.append(x)
             return lhs.func(*a)/rhs.func(*b)
-        return lhs/rhs
+        return powsimp(lhs/rhs)
 
     def as_powers_dict(self):
         d = defaultdict(int)
```
`powsimp` will combine the exponents.

**Comment by ritesh99rakesh** [comment]
> This is one way to make `match` succeed:
> 
> ```
> --- a/sympy/core/mul.py
> +++ b/sympy/core/mul.py
> @@ -1032,6 +1032,7 @@ def _combine_inverse(lhs, rhs):
>          Returns lhs/rhs, but treats arguments like symbols, so things like
>          oo/oo return 1, instead of a nan.
>          """
> +        from sympy.simplify.powsimp import powsimp
>          if lhs == rhs:
>              return S.One
>  
> @@ -1056,7 +1057,7 @@ def check(l, r):
>                  else:
>                      b.append(x)
>              return lhs.func(*a)/rhs.func(*b)
> -        return lhs/rhs
> +        return powsimp(lhs/rhs)
>  
>      def as_powers_dict(self):
>          d = defaultdict(int)
> ```
> 
> `powsimp` will combine the exponents.

Thank you, @jksuom for the insight. Sorry for not replying sooner.
I have one doubt. After making the change as suggested, I get the following output:
```
>>> a = symbols('a', postive=True)
>>> b = 1 + symbols('b', positive=True)
>>> X = LogLogistic('x', a, b)
>>> simplify(E(X))
-pi*a/((b + 1)*sin(pi*(b + 2)/(b + 1)))
```
As you can see, still `(b+1)` is not getting cancelled. Please guide me what I can do? I will open a PR in the regard once getting input from you.

**Comment by jksuom** [comment]
> (b+1) is not getting cancelled

I think that is to be expected when `b = 1 + symbols('b', positive=True)`. ( `b + 1` *is the value* of the parameter called `b`.)  Perhaps it should be defined as `b = 1 + c` where `c = symbols('c', positive=True)` to avoid confusion. Then one could substitute `b - 1` for `c` in the final result.

**Comment by ritesh99rakesh** [comment]
> > (b+1) is not getting cancelled
> 
> I think that is to be expected when `b = 1 + symbols('b', positive=True)`. ( `b + 1` _is the value_ of the parameter called `b`.) Perhaps it should be defined as `b = 1 + c` where `c = symbols('c', positive=True)` to avoid confusion. Then one could substitute `b - 1` for `c` in the final result.

Thank you @jksuom. This seems to work for me. I have made another PR #16888 since it is not related to `LogLogistic` function directly. Please review the PR #16888 and then I will work on this.

**Comment by ritesh99rakesh** [comment]
@sidhantnagpal I have added the function for the expectation since I was not getting the exact answer. Please review this.

File: sympy/stats/crv_types.py [hunk 2]
```diff
@@ -1952,6 +1954,99 @@ def Logistic(name, mu, s):
...
+    >>> cdf(X)(z)
+    1/(1 + (z/alpha)**(-beta))
+
+    >>> quantile(X)(p)
+    alpha*(p/(1 - p))**(1/beta)
+
+    References
+    ==========
+
+    .. [1] https://en.wikipedia.org/wiki/Log-logistic_distribution
```
Comment (czgdp1807) [inline sympy/stats/crv_types.py]: Wolfram or some research paper's link?

**Comment by sidhantnagpal** [comment]
I do get the desired result on your branch (without hard-coding the expectation method):
```py
In [22]: var('a, c, b', positive=True)                                          
Out[22]: (a, c, b)

In [23]: X = LogLogistic('X', a, c + 1)                                         

In [24]: E(X).subs(c, b - 1).simplify()                                         
Out[24]: 
  π⋅a   
────────
     ⎛π⎞
b⋅sin⎜─⎟
     ⎝b⎠

```
The only downside of not having the hard-coded expectation method is that the other branch of `Piecewise` doesn't look precise:
```py
In [31]: Y = LogLogistic('Y', a, b)  

In [32]: print(E(Y).simplify())                                                 
Piecewise((pi*a/(b*sin(pi/b)), 2 - 1/b > 1), (Piecewise((pi*a/(b*sin(pi/b)), 2 - 1/b > 1), (a**b*b*Integral(Y**b/(Y**(2*b) + 2*Y**b*a**b + a**(2*b)), (Y, 0, oo)), True)), True))
```

**Comment by ritesh99rakesh** [comment]
> > This is actually correct
> 
> It should ideally be undefined for `b <= 1`.
@sidhantnagpal yes, due to this reason I have coded the expectation func explicitly. What do you think should I do? Remove the expectation function?

**Comment by sidhantnagpal** [comment]
I think it is fine. This looks good to me.

**Comment by ritesh99rakesh** [comment]
@sidhantnagpal Please consider merging it.

**Pull request merged**